snuba_sdk/dsl/dsl.py

```diff
@@ -0,0 +1,296 @@
+        return expr
+
+    def visit_expr_op(self, node: Node, children: Sequence[Any]) -> Any:
+        raise InvalidQueryError("Arithmetic function not supported yet")
+        return EXPRESSION_OPERATORS[node.text]
+
+    def visit_term(self, node: Node, children: Sequence[Any]) -> Any:
+        """
+        Checks if the current node contains two term children, if so
+        then merge them into a single Function with the operator. If the
+        children are MetricQuery objects, then collapse them into a Timeseries first.
+        """
+        term, zero_or_more_others = children
+        if zero_or_more_others:
+            _, term_operator, _, coefficient, *_ = zero_or_more_others[0]
+            if isinstance(term, MetricsQuery):
+                term = self.collapse_into_timeseries(term)
+            if isinstance(coefficient, MetricsQuery):
+                coefficient = self.collapse_into_timeseries(coefficient)
+            function = Function(term_operator, [term, coefficient])
+            return MetricsQuery(query=function)
+        return term
+
+    def visit_term_op(self, node: Node, children: Sequence[Any]) -> Any:
+        raise InvalidQueryError("Arithmetic function not supported yet")
+        return TERM_OPERATORS[node.text]
+
+    def visit_coefficient(self, node: Node, children: Sequence[Any]) -> Any:
+        return children[0]
+
+    def visit_number(self, node: Node, children: Sequence[Any]) -> Any:
+        return float(node.text)
+
+    def visit_filter(self, node: Node, children: Sequence[Any]) -> Any:
+        """
+        Given a target (which could be either a MetricsQuery or Timeseries object),
+        and set its children filters and groupbys on it.
+        """
+        target, packed_filters, packed_groupbys, *_ = children
+        if not packed_filters and not packed_groupbys:
+            return target
+        assert isinstance(target, MetricsQuery) or isinstance(target, Timeseries)
+        if packed_filters:
+            _, _, first, zero_or_more_others, *_ = packed_filters[0]
+            filters = [first, *(v for _, _, _, v in zero_or_more_others)]
+            current_filters = target.filters if target.filters else []
+            filters.extend(current_filters)
+            target = target.set_filters(filters)
+        if packed_groupbys:
+            group_by = packed_groupbys[0]
+            if not isinstance(group_by, list):
+                group_by = [group_by]
+            current_groupby = target.groupby if target.groupby else []
+            group_by.extend(current_groupby)
+            target = target.set_groupby(group_by)
+        return target
+
+    def visit_condition(self, node: Node, children: Sequence[Any]) -> Any:
+        lhs, _, op, _, rhs = children
+        return Condition(lhs[0], op, rhs)
+
+    def visit_function(self, node: Node, children: Sequence[Any]) -> Any:
+        """
+        Given an target (which could be either a MetricsQuery or Timeseries object),
```

Comment:
This is confusing me a bit: why can this be anything other than a Timeseries object? A MetricsQuery could be composed of multiple Timeseries, but I don't think we want to support a MetricsQuery being composed of multiple MetricsQuerys.

Reply:
You're right, a MetricsQuery cannot be composed of multiple MetricQueries. After some more testing, it looks like target will actually only ever be a MetricsQuery. This is because if we step down the tree from the `function` node, one of its children is actually an `expression` node which resolves into a `MetricQuery`. If that's unclear, it might be easier to walk through it offline  and discuss alternatives.